### ISSUE TITLE
Communicating errors through the API

### ISSUE THREAD

**Issue by ExpHP** [dftbplus/dftbplus#249]
Currently, when DFTB+ encounters a problem with the user's input, it logs an error message and aborts the program. (sometimes even with an exit code of 0!)

This makes it difficult to use as a library, as the code using DFTB+ may have unfinished work or cleanup that it needs to perform.  It also may want to display the error message in some fashion (e.g. on its own STDERR, or in a popup notification), and perhaps even continue running.  I suspect that most larger applications (e.g. anything with a UI) will find this to be a fairly raw dealbreaker.

Are there any plans to refactor DFTB+ to propagate errors back up to the caller?

(granted, this could be very laborious and could require touching a large portion of the codebase, so I can also understand not wanting to do so...)

**Comment by aradi** [comment]
Yes, ideally DFTB+ should always return gracefully from errors. Unfortunately, this is not that straightforward in Fortran as it would be in other languages with exceptions. I would like to introduce something similar (based on some fypp-macros), but that won't happen before the current release.

**Comment by bhourahine** [comment]
Likewise, if SCC terminates abnormally, the code just cleanly stops. Instead there should be an error return or callback via the interface.

**Comment by njzjz** [comment]
I got this behavior when I use the Python API. When SCC terminates, the rest Python codes are never executed.
